File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.1_RabbitMQ/01_basic_setup/producer.py
#!/usr/bin/env python3
"""
RabbitMQ Producer - Basic Example

This script demonstrates how to:
1. Connect to RabbitMQ
2. Create a channel
3. Declare a queue
4. Publish messages to the queue
"""

import pika
import sys
import json
from datetime import datetime


def create_connection():
    """Create a connection to RabbitMQ"""
    credentials = pika.PlainCredentials('admin', 'admin123')
    parameters = pika.ConnectionParameters(
        host='localhost',
        port=5672,
        virtual_host='/',
        credentials=credentials,
        heartbeat=600,
        blocked_connection_timeout=300
    )

    try:
        connection = pika.BlockingConnection(parameters)
        print("✓ Connected to RabbitMQ")
        return connection
    except pika.exceptions.AMQPConnectionError as e:
        print(f"✗ Failed to connect to RabbitMQ: {e}")
        sys.exit(1)


def send_message(channel, queue_name, message):
    """Send a message to the queue"""
    # Declare the queue (idempotent - won't recreate if exists)
    channel.queue_declare(
        queue=queue_name,
        durable=True,  # Queue survives broker restart
        exclusive=False,  # Queue can be accessed by other connections
        auto_delete=False  # Queue won't be deleted when last consumer unsubscribes
    )

    # Publish the message
    channel.basic_publish(
        exchange='',  # Default exchange
        routing_key=queue_name,  # Queue name
        body=message,
        properties=pika.BasicProperties(
            delivery_mode=2,  # Make message persistent
            content_type='application/json',
            timestamp=int(datetime.now().timestamp())
        )
    )

    print(f"✓ Sent message to queue '{queue_name}': {message}")


def main():
    """Main function"""
    queue_name = 'hello_queue'

    # Create connection and channel
    connection = create_connection()
    channel = connection.channel()
    print("✓ Channel created")

    # Send multiple messages
    messages = [
        {"id": 1, "text": "Hello World!", "timestamp": datetime.now().isoformat()},
        {"id": 2, "text": "Hello RabbitMQ!", "timestamp": datetime.now().isoformat()},
        {"id": 3, "text": "First message queue tutorial", "timestamp": datetime.now().isoformat()},
    ]

    for msg in messages:
        message_json = json.dumps(msg)
        send_message(channel, queue_name, message_json)

    # Close the connection
    connection.close()
    print("✓ Connection closed")
    print(f"\n📊 Summary: Sent {len(messages)} messages to queue '{queue_name}'")


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('\n✗ Interrupted by user')
        sys.exit(0)
    except Exception as e:
        print(f'✗ Error: {e}')
        sys.exit(1)

File: 5_HandsOn_Tutorials/5.8_Message_Queues/5.8.1_RabbitMQ/01_basic_setup/producer.ts
#!/usr/bin/env ts-node
import * as amqp from 'amqplib';
import { Connection, Channel } from 'amqplib';

/**
 * Message interface for type safety
 */
interface HelloMessage {
    id: number;
    text: string;
    timestamp: string;
}

/**
 * Send messages to RabbitMQ queue
 */
async function sendMessages(): Promise<void> {
    let connection: Connection | null = null;
    let channel: Channel | null = null;

    try {
        // Connection parameters
        const connectionUrl = 'amqp://admin:password@localhost:5672/';

        // Establish connection
        connection = await amqp.connect(connectionUrl);
        channel = await connection.createChannel();

        // Declare queue (idempotent - safe to call multiple times)
        const queueName = 'hello';
        await channel.assertQueue(queueName, {
            durable: true,  // Survive broker restart
            exclusive: false,
            autoDelete: false
        });

        // Send 5 messages
        for (let i = 1; i <= 5; i++) {
            const message: HelloMessage = {
                id: i,
                text: `Hello World #${i}`,
                timestamp: new Date().toISOString()
            };

            const messageBuffer = Buffer.from(JSON.stringify(message));

            channel.sendToQueue(queueName, messageBuffer, {
                persistent: true,  // Make message persistent
                contentType: 'application/json'
            });

            console.log(`✓ Sent: ${JSON.stringify(message)}`);

            // Wait 1 second between messages
            await new Promise(resolve => setTimeout(resolve, 1000));
        }

        console.log('\n✓ All messages sent successfully');

    } catch (error) {
        console.error('Error sending messages:', error);
        throw error;
    } finally {
        // Clean up resources
        try {
            if (channel) await channel.close();
            if (connection) await connection.close();
        } catch (closeError) {
            console.error('Error closing connection:', closeError);
        }
    }
}

// Run the function
if (require.main === module) {
    sendMessages().catch(error => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}

export { sendMessages, HelloMessage };
